### ISSUE TITLE
Resource for individual animal observation statistics

### ISSUE THREAD

**Issue by JeroenW1** [adewg/ICAR#465]
As herd monitoring system manufacturer we need an option to facilitate the transfer of individual-level animal measurements. Together with some of our herd management system integration partners we came up with the following proposal for a new animal measurement endpoint. Our goal is to maintain the versatility of the animal measurements data structure, ensuring its compatibility with a diverse array of measurement types collected by various systems and sensors. It is essential that this endpoint is designed with foresight, eliminating the need for new endpoints with the introduction of novel measurement types.

Initially, we aim to utilize this animal measurements endpoint to export data concerning the daily time spent on behaviours like ruminating, eating, and inactivity. Later on we would also like to use the endpoint for exports of data such as heat activity values, health score values, automatically measured locomotion scores and other time series datasets.

# Proposal
## Animal Measurements endpoint
This endpoint returns data about animal measurements for all animals on a specific location. 

GET: /v1/locations/{location-scheme}/{location-id}/animal-measurements

#### Path
-	location-scheme: location identifier scheme
-	location-id: id

#### Query 
-	measurement-types: List of measurement types that you want to request, see measurement types Enum below
-	measurement-datetime-from: The start datetime of the range in which you want to request data (UTC)
-	measurement-datetime-to: The end datetime of the range in which you want to request data (UTC)
-	page-size: The number of items to display on one page of results. 
-	page: The page number
-	animal-scheme: Select the animal scheme
-	animal-id: Select a specific animal

### Response Schema

- **animalMeasurements** [list]
  - **animal**
    - scheme [string]: Identfies the data source as com.manufacturer.animal.id
    - id [string]: The unique Manufacturer animal identifier (UUID)
  - **measurements** [list]
    - **type** [enum]: The measurement type
      - DailyRuminationTime
      - DailyEatingTime
      - DailyInactiveTime
      - Last24HourRuminationTime
      - Last24HourEatingTime
      - Last24HourInactiveTime
    - **value**: The measurement value
    - **valueType** [enum]: the type of the measurement values
      - Number
    - **unit** [enum]: the unit of the measurement, UN/CEFACT Common Code for Units of Measurement.
      - SEC
      - MIN
    - **isComplete** [boolean] (optional): Flag that indicates if the measurement is complete. For daily measurements this flag is complete if the value is based on a full day of measurements. If not all measurements have been received the value will be false.
    - **dateTime** [datetime]: The start date and time of the measurement in UTC.

### Response example

```json
{
  "view": {
    "totalItems": 0,
    "totalPages": 0,
    "pageSize": 0,
    "currentPage": 0
  },
  "member": [
    {
      "resourceType": "string",
      "self": "string",
      "meta": {
        "source": "string",
        "sourceId": "string",
        "isDeleted": true,
        "modified": "2024-02-07T22:14:07.635Z",
        "created": "2024-02-07T22:14:07.635Z",
        "creator": "string",
        "validFrom": "2024-02-07T22:14:07.635Z",
        "validTo": "2024-02-07T22:14:07.635Z"
      },
      "location": {
        "scheme": "com.nedap.farm.id",
        "id": "2f7c3d10-4d2c-4758-81b4-4d8f496ea659"
      },
      "animalMeasurements": [
        {
          "animal": {
            "scheme": "com.nedap.animal.id",
            "id": "fd1b3d96-2620-4040-a1a2-8d913b11b788"
          },
          "measurements": [
            {
              "type": "DailyRuminationTime",
              "value": 400,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-07T23:00:00.000Z"
            },
            {
              "type": "DailyEatingTime",
              "value": 300,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-07T23:00:00.000Z"
            },
            {
              "type": "DailyInactiveTime",
              "value": 500,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-07T23:00:00.000Z"
            },
            {
              "type": "Last24HourRuminationTime",
              "value": 350,
              "valueType": "Number",
              "unit": "MIN",
              "isComplete": true,
              "dateTime": "2024-02-08T16:14:07.635Z"
            }
          ]
        }
      ]
    }
  ]
}
```

**Comment by cookeac** [comment]
Some suggestions:
* Think about the event name and intent -- is this about time-based activities (in which case `ActivityObservation` might be more appropriate) or could it represent all measurements (body condition score, chest measurement...)
* Consider whether an enum is appropriate for the type of measurement, or whether a scheme and ID (ideally published in `well-known` identifiers) might be used for flexibility. See `icarCarcassObservationsEventResource` with its array of `icarCarcassObservationType.
* For the measures themselves, could we generalise `icarMassMeasureType` (or at least use the same pattern)?
* Consider identifying the time period covered. If this is (as it should be) an `icarAnimalEventCoreResource` derivative then we already have an event date time (i.e. time these metrics were generated) - would we just identify the start time? What does the `dateTime` in each measure mean?

**And one from the chat on 2024-05-08**
Do  we know how the "Last 24H" vs "daily" entries work for measurements? Would these be posted once-per-day-per-animal or multiple times per day? If multiple per day, would the daily be the current accumulation since previous midnight and the 24H be a rolling 24H window at each update time? 
Given the explanation for `isComplete` this seems to be the case, but just thinking ahead to how we might need to query/aggregate these and how many updates we might be getting per day per animal.

**Comment by cookeac** [comment]
At our meetings today we agreed:
- Rename the event to something like icarActivitySummaryResource
- Include ($ref) icarResource rather than icarAnimalEventCoreResource and add the necessary animal identifiers. This avoids confusion that it is an event observation when it is actually a requested summary.
- Make it clear that the daily aggregation in the request is carried out using in LOCAL date time for the location not UTC, but keep all date times in UTC time.
- Expand the set of metric types (suggested - "metric" rather than "type") to a superset list of activities. For instance, Lying, Standing, Resting (Inactive?), Eating, Drinking, Ruminating.
- Rename isComplete to isIncomplete (or InsufficientData) so that it only has to be provided when the data is not sufficiently complete to return useful values, but it is helpful to indicate that the source system noted the request for the data.

**Comment by cookeac** [comment]
Question from meeting 2024-07-10 
What about other "activity" data that is not about behaviour? For instance rumen pH or other metrics?

**Comment by jdoemel** [comment]
@JeroenW1 @cookeac is this moving forward?  If so, do we have a timeframe.  VAS would like to get this implemented. However, I don't want to do that until ICAR has a resource that we can follow.

**Comment by cookeac** [comment]
Discussed at the North America time zone meeting on 2024-10-02

- Reviewed PR #486 
- Noted when reading the JSON schema itself, its a little confusing to have some details in `icarStatisticsType` by reference, and other attributes in `icarObservationStatisticsType`, but when using this in tools such as Swagger it will all be fine.
- Noted that we need to add a further value for the measurement value, and distinguish between this and the aggregation value. 
 
For instance, a behaviour such as lying might just have an aggregation value (`icarStatisticsType.value`) with `units` in "MIN". However, an aggregation of rumen pH might have an aggregation value in "MIN" for a given **measurement value** of 5.5 (i.e. the cow spent x minutes in the last 24 hours with a rumen pH of 5.5).

- Discussed API semantics. The caller will specify the location or specific animal, and the date range for results.
- The caller should be able to specify the aggregation time scales required (e.g. aggregate to 24 hour days or 96 hour periods), and the service should return a 4xx error if it can't do that aggregation (422 unprocessable?). What about multiple time periods - e.g. working days AND last 24 hours, or should you only ask for one aggregation?
- Should the caller be able to specify what metrics are required? A metric is identified by a scheme and ID. Should you only be able to ask for all metrics (no filter) or just one metric (single metric filter)? Could there be categories of metrics?

Discussed other use cases. The examples so far have been behaviour time, or time at a given rumen pH, or steps in a day, but you could have total litres or kg of milk produced in a day, or number of milking visits... Any individual animal statistic of aggregation.
- We will need to put effort into appropriate documentation so the two items - measurement value and aggregation value - are clear.

**Comment by cookeac** [comment]
Discussed at the European morning meeting on 2024-10-03

- Agreed to rename the title of this issue to be a resource for statistics of individual animal observations
- Decided that we did not need a separate measurement value to aggregation value. The case of knowing the amount of time spent at a given rumen pH value is just an example and does not make sense in the real world.
- Decided that its important that the metric name or description should clarify the sort of measurement and aggregation expected. For instance, LyingTime tells us that the value and units will be time-based.
- Decided that we did need an aggregation called "Range" being the difference between min and max values in a period.
- Decided that we did need an aggregation called "Score" meaning a computed index (e.g. health index, health score, locomotion score). A score does not have a set of units (units are not required).
- We discussed units for things that are counts (for instance, the number of milking visits or the number of steps). Is there a UN/CEFACT unit for these?

_After the meeting I researched. The UN/CEFACT unit for "number of individual items" is "EA" (each). The UN/CEFACT for the "number of actions" is "ACT". So the number of milking visits could be specified without units, or by using "ACT"._

We discussed API semantics:
- Agreed that the date range is important - from and to
- Agreed that animal (scheme and id) is optional - to filter to a specific animal rather than all animals at the location
- Agreed that metric (scheme and id) is optional - to filter to a specific metric rather than all metrics aggregated by the system
- Agreed that the duration for aggregation must be specified (e.g. 1D, 24H, 96H etc). 

Finally, we noted that "96H" should be added to `icarDurationType`, that `location` should be removed from the resource as it is already an attribute of `icarResource`, and that `dateFrom` and `dateTo` should also be removed from the resource as they duplicate the parameters, will be the same for each resource in the response, and don't add useful data when saved,

**Comment by cookeac** [comment]
Resolved by #486
